experiment/selinux/src/selinux.rs

```diff
@@ -0,0 +1,302 @@
+    let label = format!("/proc/{}/attr/current", pid);
+    return read_con(label.as_str());
+}
+
+pub fn exec_label() -> Result<String, std::io::Error> {
+    return read_con(attr_path("exec").as_str());
+}
+
+pub fn set_exec_label(label: &str) {
+    panic!("not implemented yet")
+}
+
+pub fn set_task_label(label: &str) {
+    panic!("not implemented yet")
+}
+
+pub fn set_socket_label(label: &str) {
+    panic!("not implemented yet")
+}
+
+pub fn socket_label() {
+    panic!("not implemented yet")
+}
+
+pub fn peer_label() {
+    panic!("not implemented yet")
+}
+
+pub fn set_key_label(label: &str) -> Result<(), std::io::Error> {
+    match write_con("/proc/self/attr/keycreate", label) {
+        Ok(v) => return Ok(v),
+        //TODO: update error
+        Err(e) => return Err(e),
+    }
+}
+
+pub fn key_label() {
+    panic!("not implemented yet")
+}
+
+pub fn clear_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn reserve_label(label: &str) {
+    panic!("not implemented yet")
+}
+
+pub fn ro_file_label() {
+    panic!("not implemented yet")
+}
+
+pub fn kvm_container_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn init_container_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn container_labels() {
+    panic!("not implemented yet")
+}
+
+pub fn priv_container_mount_label() {
+    panic!("not implemented yet")
+}
+
+pub fn format_mount_label(src: &str, mount_label: &str) -> String {
+    return format_mount_label_by_type(src, mount_label, "context")
+}
+
+pub fn format_mount_label_by_type(src: &str, mount_label: &str, context_type: &str) -> String {
+    let mut formatted_src = src.to_owned();
+
+    if !mount_label.is_empty() {
+        if formatted_src.is_empty() {
+            formatted_src = format!("{}=\"{}\"", context_type, mount_label);
+        } else {
+            formatted_src = format!("{},{}=\"{}\"", formatted_src, context_type, mount_label);
+        }
+    }
+    return formatted_src
+}
+
+pub fn write_con(fpath: &str, val: &str) -> Result<(), std::io::Error> {
+    panic!("not implemented yet");
+}
+
+pub fn is_proc_handle(file: &File) -> Result<(), std::io::Error> {
+    loop {
+        match statfs::fstatfs(file.as_fd()) {
+            Ok(stat) if stat.filesystem_type() == statfs::PROC_SUPER_MAGIC => break,
+            Ok(_) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other, format!("file {} is not on procfs", file.as_raw_fd())
+                ));
+            },
+            Err(Errno::EINTR) => continue,
+            Err(err) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("fstatfs failed: {}", err)))
+            }
+        }
+    }
+    Ok(())
+}
+
+pub fn read_con_fd(file: &mut File) -> Result<String, std::io::Error> {
+    let mut data = String::new();
+    file.read_to_string(&mut data)?;
+
+    let trimmed_data = data.trim_end_matches(char::from(0));
```

Comment:
Is this necessary? If I'm correct, then `read_to_string` returns a Rust String, which should not be null terminated in first place. Can the file content itself have a null byte in it which we are trying to remove here?

Reply:
I added test cases in `test_read_con_fd`.
Seems that  trimming is necessary if we want to remove null bytes on the end of a file.
Does it make sense?
Sorry if I misunderstood the context or something

Reply:
Hey, I think this is ok for now, we can re-visit this in later PRs if needed. I was just wondering that if the `read_to_string` fn itself will remove the trailing `0` s, ie. would `test_read_con_fd` pass even if we remove this trim call.